### ISSUE TITLE
pp_sys.c:1580: incompatible type for argument 5 of `recvfrom'

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4562]
Migrated from [rt.perl.org#7885](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=7885) (status was 'resolved')

Searchable as RT7885$

**Comment by p5pRT** [comment]
### From ask@wisdom.weizmann.ac.il

I use gcc version 2\.96 20000731 \(Red Hat Linux 7\.1 2\.96\-85\) to build
bleadperl 12887 and got&#8203;:

pp\_sys\.c&#8203;: In function \`Perl\_pp\_sysread'&#8203;:
pp\_sys\.c&#8203;:1580&#8203;: incompatible type for argument 5 of \`recvfrom'

In pp\_sys\.i we have the following&#8203;:
\# 87 "/usr/include/sys/socket\.h" 3
typedef union \{ struct sockaddr \*\_\_restrict \_\_sockaddr\_\_; struct
sockaddr\_at \*\_\_restrict \_\_sockaddr\_at\_\_; struct sockaddr\_ax25 \*\_\_restrict
\_\_sockaddr\_ax25\_\_; struct sockaddr\_dl \*\_\_restrict \_\_sockaddr\_dl\_\_; struct
sockaddr\_eon \*\_\_restrict \_\_sockaddr\_eon\_\_; struct sockaddr\_in \*\_\_restrict
\_\_sockaddr\_in\_\_; struct sockaddr\_in6 \*\_\_restrict \_\_sockaddr\_in6\_\_; struct
sockaddr\_inarp \*\_\_restrict \_\_sockaddr\_inarp\_\_; struct sockaddr\_ipx
\*\_\_restrict \_\_sockaddr\_ipx\_\_; struct sockaddr\_iso \*\_\_restrict
\_\_sockaddr\_iso\_\_; struct sockaddr\_ns \*\_\_restrict \_\_sockaddr\_ns\_\_; struct
sockaddr\_un \*\_\_restrict \_\_sockaddr\_un\_\_; struct sockaddr\_x25 \*\_\_restrict
\_\_sockaddr\_x25\_\_;\} \_\_SOCKADDR\_ARG ;

extern int recvfrom \(int \_\_fd\, void \*\_\_restrict \_\_buf\, size\_t \_\_n\, int
\_\_flags\, \_\_SOCKADDR\_ARG \_\_addr\, socklen\_t \*\_\_restrict \_\_addr\_len\);

count = recvfrom\(PerlIO\_fileno\(\(\(XPVIO\*\) \(io\)\->sv\_any\)\->xio\_ifp\)\, buffer\,
length\, offset\, \(struct sockaddr \*\)namebuf\, &bufsize\);

The patch
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
\*\*\* pp\_sys\.c~    Sat Oct 20 00&#8203;:13&#8203;:23 2001
\-\-\- pp\_sys\.c    Thu Nov  8 10&#8203;:32&#8203;:06 2001
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
\*\*\* 1580\,1582 \*\*\*\*
&nbsp;      count = PerlSock\_recvfrom\(PerlIO\_fileno\(IoIFP\(io\)\)\, buffer\, length\, offset\,
\!               \(struct sockaddr \*\)namebuf\, &bufsize\);
&nbsp;      if \(count \< 0\)
\-\-\- 1580\,1582 \-\-\-\-
&nbsp;      count = PerlSock\_recvfrom\(PerlIO\_fileno\(IoIFP\(io\)\)\, buffer\, length\, offset\,
\!               \(\_\_SOCKADDR\_ARG\)\(struct sockaddr \*\)namebuf\, &bufsize\);
&nbsp;      if \(count \< 0\)
\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<

solves the problem\, but there are many other instances of the problem and
I doubt if it is the best way to solve it\.

Regards\,
ASK

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

recvfrom\(2\) is defined to take a struct sockaddr \*\, and it isn't Perl's
fault if something prevents that from working correctly\.

In any case\, \_\_SOCKADDR\_ARG\, quite apart from being glibc\-specific\, is
not intended to be user\-visible \(look at the hideously complicated bits
near the top of sys/socket\.h which \#define it\)\.

\- ams

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

OK\, let's look on recvfrom\(2\)&#8203;:
&nbsp;       \#include \<sys/types\.h>
&nbsp;       \#include \<sys/socket\.h>

&nbsp;       int  recvfrom\(int  s\,  void  \*buf\,  size\_t len\, int flags\,
&nbsp;       struct sockaddr \*from\, socklen\_t \*fromlen\);

And the funny thing is that the program a\.c

\#include \<sys/types\.h>
\#include \<sys/socket\.h>
int main\(\)\{
&nbsp;  int s;
&nbsp;  void \*buf;
&nbsp;  size\_t len;
&nbsp;  int flags;
&nbsp;  struct sockaddr \*from;
&nbsp;  socklen\_t \*fromlen;
&nbsp;  recvfrom\(s\, buf\, len\, flags\, from\, fromlen\);
\}

is compiled \(\`sh  cflags "optimize='\-g \-O2'" libperl\.so pp\_sys\.o\` \-fpic
a\.c\) without problems \(but with warnings\) and even run without core
dump :\-\) The seriuos thing is that a\.i contains

extern int recvfrom \(int \_\_fd\, void \*\_\_restrict \_\_buf\, size\_t \_\_n\, int
\_\_flags\, struct sockaddr \*\_\_restrict \_\_addr\, socklen\_t \*\_\_restrict
\_\_addr\_len\);

which is differ from whatever we see in pp\_sys\.c\. The interesting thing is
that the patch
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
\*\*\* pp\_sys\.c~   Sat Oct 20 00&#8203;:13&#8203;:23 2001
\-\-\- pp\_sys\.c    Thu Nov  8 15&#8203;:10&#8203;:57 2001
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
\*\*\* 16\,17 \*\*\*\*
\-\-\- 16\,20 \-\-\-\-

\+ \#include \<sys/types\.h>
\+ \#include \<sys/socket\.h>
\+
&nbsp;  \#include "EXTERN\.h"
\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<
solves the problem\, but if we add the includes after "perl\.h" the problem
remains\. So\, it is looks like "perl\.h" should be patched somehow\.

Regards\,
ASK

**Comment by p5pRT** [comment]
### From @doughera88

> which is differ from whatever we see in pp\_sys\.c\. The interesting thing is
> that the patch
> >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> \*\*\* pp\_sys\.c~   Sat Oct 20 00&#8203;:13&#8203;:23 2001
> \-\-\- pp\_sys\.c    Thu Nov  8 15&#8203;:10&#8203;:57 2001
> \*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
> \*\*\* 16\,17 \*\*\*\*
> \-\-\- 16\,20 \-\-\-\-
> 
> \+ \#include \<sys/types\.h>
> \+ \#include \<sys/socket\.h>
> \+
>   \#include "EXTERN\.h"
> \<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<
> solves the problem\, but if we add the includes after "perl\.h" the problem
> remains\. So\, it is looks like "perl\.h" should be patched somehow\.

Hmm\.  It's very hard to guess what's going on here\.  Certainly perl\.h
should be including \<sys/types\.h> and \<sys/socket\.h>\.  Could you verify
that they are getting included there?  If so\, apparently there's something
sensitive to the order of \#includes in perl\.h\.

Hmm\.  A wild guess&#8203;:  In perl\.h\, could you try moving the \<netinet/in\.h>
and \<arpa/inet\.h> includes to below the \<sys/socket\.h> include?

Finally\, could you please send along the output of the \./myconfig script?
I'm afraid I don't know anywhere near enough about RedHat products to
guess which libc and kernel you are using\.

\-\- 
&nbsp;    Andy Dougherty        doughera@&#8203;lafayette\.edu
&nbsp;    Dept\. of Physics
&nbsp;    Lafayette College\, Easton PA 18042

**Comment by p5pRT** [comment]
### From @doughera88

> The investigation shows more details\. I have 2 gcc&#8203;:
> gcc version 3\.1 20011031 \(experimental\)
> gcc version 2\.96 20000731 \(Red Hat Linux 7\.1 2\.96\-85\)
> 
> \`sh Configure \-des \-Dusedevel && make' works OK with 2\.96\, but
> not with 3\.1\.
> 

> Now I think I know the real reason&#8203;: program for testing attributes support
> which is compiled in Configure&#8203;:8716
> 
> cc \-fno\-strict\-aliasing \-I/usr/local/include \-D\_LARGEFILE\_SOURCE
> \-D\_FILE\_OFFSET\_BITS=64 \-c attrib\.c
> 
> raise warnings
> 
> cc1&#8203;: warning&#8203;: changing search order for system directory
> "/usr/local/include"
> cc1&#8203;: warning&#8203;:   as it has already been specified as a non\-system directory
> 
> and so Configure thinks "Your C compiler doesn't fully support
> \_\_attribute\_\_\."

Good detective work\.  Thanks\. \(Why gcc\-3\.1 is warning about a very very
very common practice that has worked for years and years and years is a
completely different matter\, of course\, but one I guess we'll have to work
around\.\)

\-\- 
&nbsp;    Andy Dougherty        doughera@&#8203;lafayette\.edu
&nbsp;    Dept\. of Physics
&nbsp;    Lafayette College\, Easton PA 18042

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

> Hmm\.  A wild guess&#8203;:  In perl\.h\, could you try moving the \<netinet/in\.h>
> and \<arpa/inet\.h> includes to below the \<sys/socket\.h> include?
No improovements\.

> Finally\, could you please send along the output of the \./myconfig script?
> I'm afraid I don't know anywhere near enough about RedHat products to
> guess which libc and kernel you are using\.
The investigation shows more details\. I have 2 gcc&#8203;:
gcc version 3\.1 20011031 \(experimental\)
gcc version 2\.96 20000731 \(Red Hat Linux 7\.1 2\.96\-85\)

\`sh Configure \-des \-Dusedevel && make' works OK with 2\.96\, but
not with 3\.1\.

myconfig for 3\.1 gives&#8203;:
Summary of my perl5 \(revision 5\.0 version 7 subversion 2 patch 12910\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.4\.2\-2\, archname=i686\-linux
&nbsp;    uname='linux verdi\.wisdom\.weizmann\.ac\.il 2\.4\.2\-2 \#1 sun apr 8 20&#8203;:41&#8203;:30 edt 2001 i686 unknown '
&nbsp;    config\_args='\-des \-Dusedevel'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=define
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-fno\-strict\-aliasing \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2'\,
&nbsp;    cppflags='\-fno\-strict\-aliasing \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='3\.1 20011031 \(experimental\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-lndbm \-lgdbm \-ldl \-lm \-lc \-lcrypt \-lutil
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lc \-lcrypt \-lutil
&nbsp;    libc=/lib/libc\-2\.2\.2\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic'
&nbsp;    cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'

I compare two versions\, and the most notable difference is
\*\*\* bleadperl\.296/config\.h      Fri Nov  9 14&#8203;:31&#8203;:45 2001
\-\-\- bleadperl/config\.h  Fri Nov  9 14&#8203;:32&#8203;:37 2001
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
\*\*\* 39 \*\*\*\*
\! \#define HASATTRIBUTE  /\*\*/
\-\-\- 39 \-\-\-\-
\! /\*\#define HASATTRIBUTE        / \*\*/

In other words perl think that gcc 3\.1 couldn't check for function
attributes\, but 2\.96 could\. After I uncoment it everything is compiled OK
with 3\.1 \(98\.97% tests passed\)

Now I think I know the real reason&#8203;: program for testing attributes support
which is compiled in Configure&#8203;:8716

cc \-fno\-strict\-aliasing \-I/usr/local/include \-D\_LARGEFILE\_SOURCE
\-D\_FILE\_OFFSET\_BITS=64 \-c attrib\.c

raise warnings

cc1&#8203;: warning&#8203;: changing search order for system directory
"/usr/local/include"
cc1&#8203;: warning&#8203;:   as it has already been specified as a non\-system directory

and so Configure thinks "Your C compiler doesn't fully support
\_\_attribute\_\_\."

Regards\,
ASK

**Comment by p5pRT** [comment]
### From @doughera88

perl\-5\.8\.0 should now work with gcc\-3\.1\.  The redundant \-I flags
are now removed\.

**Comment by p5pRT** [comment]
@doughera88 - Status changed from 'open' to 'resolved'
